### ISSUE TITLE
Unable to install istio on EKS with IPv6

### ISSUE THREAD

**Issue by baalajimaestro** [istio/istio#36961]
### Bug Description

Tried my best to install istio on EKS, and everytime `istio-ingressgateway` pod fails to mark as ready, with readiness probe failing with an error log mentioning `Readiness probe failed: Get "http://[IPv6 endpoint of the pod]:15021/healthz/ready": dial tcp [same IPv6 endpoint]:15021: connect: connection refused`. Istioctl keeps waiting for 5 minutes and fails to spin up istio. Is there any IPv6-specific configuration that I have to include to ensure this works?

### Version

```prose
client version: 1.12.2
control plane version: 1.12.2
data plane version: 1.12.2 (1 proxies)

Client Version: v1.23.2
Server Version: v1.21.5-eks-bc4871b
WARNING: version difference between client (1.23) and server (1.21) exceeds the supported minor version skew of +/-1

Didnt use helm
```

### Additional Information

Target cluster context: devops@CLUSTER_NAME.ap-south-1.eksctl.io

Running with the following config:

istio-namespace: istio-system
full-secrets: false
timeout (mins): 30
include: {  }
exclude: { Namespaces: kube-system,kube-public,kube-node-lease,local-path-storage }
end-time: 2022-01-24 16:45:47.136144463 +0530 IST

Cluster endpoint: https://HIDDEN.gr7.ap-south-1.eks.amazonaws.com
CLI version:
version.BuildInfo{Version:"1.12.2", GitRevision:"af0d66fd0aa363e9a7b0164f3a94ba36252fe60f", GolangVersion:"go1.17.5", BuildStatus:"Clean", GitTag:"1.12.2"}

The following Istio control plane revisions/versions were found in the cluster:
Revision default:
&version.MeshInfo{
    {
        Component: "pilot",
        Info:      version.BuildInfo{Version:"1.12.2", GitRevision:"af0d66fd0aa363e9a7b0164f3a94ba36252fe60f-dirty", GolangVersion:"", BuildStatus:"Modified", GitTag:"1.12.2"},
    },
}

The following proxy revisions/versions were found in the cluster:
Revision default: Versions {1.12.2}

Fetching proxy logs for the following containers:

istio-system/istio-ingressgateway/istio-ingressgateway-66cc5dcb7d-tgfwf/istio-proxy
istio-system/istiod/istiod-577b69955f-nx62b/discovery

Fetching Istio control plane information from cluster.

Running istio analyze on all namespaces and report as below:
Analysis Report:
Info [IST0102] (Namespace cert-manager) The namespace is not enabled for Istio injection. Run 'kubectl label namespace cert-manager istio-injection=enabled' to enable it, or 'kubectl label namespace cert-manager istio-injection=disabled' to explicitly mark it as not needing injection.
Info [IST0102] (Namespace default) The namespace is not enabled for Istio injection. Run 'kubectl label namespace default istio-injection=enabled' to enable it, or 'kubectl label namespace default istio-injection=disabled' to explicitly mark it as not needing injection.
Creating an archive at /home/baalajimaestro/bug-report.tar.gz.
Cleaning up temporary files in /tmp/bug-report.
Done.

**Comment by bianpengyuan** [comment]
Could you get the log from the istio ingress pod?

**Comment by baalajimaestro** [comment]
https://paste.baalajimaestro.me/eL16ke

Logs from istio-ingress pod

Looking in, i notice envoy command having local-ip-address v4, is that the issue here?

**Comment by bianpengyuan** [comment]
Looks like if pod have at least one ipv4 address, the envoy proxy will be configured to use v4 for local address: https://github.com/istio/istio/blob/026e119a9abedd973b5efb5a95c64204069e8153/pilot/pkg/util/network/ip.go#L146 Although I don't think this behavior is desired, not only kubelet uses the first pod ip for prober, but also the proxy will use the first ip when getting config from the control plane. @howardjohn do you know if this is intentional?

**Comment by howardjohn** [comment]
Dual stack is not support in way at all

**Comment by bianpengyuan** [comment]
> Dual stack is not support in way at all

Yeah Although I think we should make the behavior consistent at Envoy configuration and agent status listener? IIUC, pilot uses first ip to decide proxy ip version, but agent uses any presence of ipv4 to decide ip version it listens on, which does not seem right to me.

**Comment by howardjohn** [comment]
Hm..
https://github.com/istio/istio/commit/e997c2230b65c3a54ab64a6a208dd3aafb13e211
(see commit comment) shows we have quite a bit of inconsistency. One
problem is in envoy config we have 2 bools - support ipv6 and support ipv4.
But in these cases we need to pick exactly 1?

**Comment by bianpengyuan** [comment]
Actually looks like all listeners (including static status listener) are built based on `support ipv4/ipv6`, and ipv4 takes precedence. But endpoint just uses the first ip in the list.

However at kubelet, it will just pick the first at the podIPs list, which does not match the listener behavior. @baalajimaestro I think we don't have a good way to solve this at the moment, since dual stack is not well supported at Istio. You can work around for now by use only IPv6 or IPv4 for your cluster.

**Comment by baalajimaestro** [comment]
The cluster is IPv6 only, and IPs are being only IPv6 type inside the cluster. The IP provided to the pod is, however a part of the VPC, and not a part of the cluster as such.

The IP provided to the pod starts with `2406:xxx` while the one thats for the service starts with `fdc2:xxx`

The loadbalancer attached to this is dualstack type but the cluster isnt. The IPv6 support on EKS was added quite recently, so there isnt much guides on how to implement it properly.

The Pod:

```
Name:         istio-ingressgateway-66cc5dcb7d-f4mzq
Namespace:    istio-system
Priority:     0
Node:         ip-192-168-13-78.ap-south-1.compute.internal/2406:xxxx
Start Time:   Wed, 26 Jan 2022 13:37:16 +0530
Labels:       app=istio-ingressgateway
              chart=gateways
              heritage=Tiller
              install.operator.istio.io/owning-resource=unknown
              istio=ingressgateway
              istio.io/rev=default
              operator.istio.io/component=IngressGateways
              pod-template-hash=66cc5dcb7d
              release=istio
              service.istio.io/canonical-name=istio-ingressgateway
              service.istio.io/canonical-revision=latest
              sidecar.istio.io/inject=false
Annotations:  kubernetes.io/psp: eks.privileged
              prometheus.io/path: /stats/prometheus
              prometheus.io/port: 15020
              prometheus.io/scrape: true
              sidecar.istio.io/inject: false
Status:       Running
IP:           2406:xxx
IPs:
  IP:           2406:xxx
```

This fails with

```
Warning  Unhealthy  3m24s (x22 over 4m5s)  kubelet            Readiness probe failed: Get "http://[IP same as pod]:15021/healthz/ready": dial tcp [IP same as pod]:15021: connect: connection refused
```

Another interesting thing i noted is:

This IP provided somehow works for istiod for the readiness, since its marked as ready and didnt fail the readiness check.

Will be happy to provide more logs to debug this issue.

**Comment by mKeRix** [comment]
IPv6 in EKS with the aws-cni will add a host-local NAT for IPv4 traffic by default. This is done through a new plugin in the CNI, which maps a private IPv4 address to each pod (the one starting with `169`). The address is only unique to the underlying host and will not be advertised to Kubernetes anywhere.

You can disable this behavior and make it a pure IPv6 cluster by disabling the plugin in the CNI config - this will cause you to lose IPv4 access to resources in the private subnets however. Depending on regional availability of NAT64 and DNS64 you may lose IPv4 access to the Internet as well.

**Comment by bianpengyuan** [comment]
@mKeRix do you happen to know if it is possible to distinguish the host-local ipv4 programmatically? If we can reliably tell that the address is not going to be advertised to kubernetes, we can exclude it from the pod ip searching by default: https://github.com/istio/istio/blob/69bc1e32bc7b79e79f18a94486edfbb3dc9b4983/pilot/pkg/util/network/ip.go#L61

**Comment by mKeRix** [comment]
For the EKS implementation specifically the easiest way to filter would be to do either of the following:

- IP is in the 169.254.172.0/22 range (to be more generic 169.254.0.0/16 could be used, which are all link-local addresses)
- filter the `v4if0` interface 

If you want to base it on what is advertised to K8s the only way I can think of doing this within a pod is setting an environment variable to `.status.podIPs`.

Personally, I think excluding the link-local address range seems like the best solution.

**Comment by mKeRix** [comment]
I've been tinkering with the Istio code today and got a custom build of mine to work on EKS IPv6 clusters. I went the route of excluding the link-local addresses from detection by using `IsLinkLocalUnicast` as a check.

So far I did discover an additional caveat: Istio will also configure IPv4 redirections, but those connections will be dropped. This is relevant because there may actually be IPv4 traffic, even if the cluster is single-stack - the host-local NAT should continue to work. I presume that this happens because the proxy will only listen on a single IP family, so there is nothing there to accept the outbound traffic in the proxy image. You can work around this by specifying `excludeIPRanges: 0.0.0.0/0` in your config (or the respective annotation on the pods). That means the IPv4 traffic is unrestricted though of course.

**Comment by baalajimaestro** [comment]
Hi @mKeRix can you help me with the config, and is the custom build having anything relevant to IPv6 support?

**Comment by mKeRix** [comment]
The custom build is required to get it working - essentially I cherry-picked the changes from #37164 on top of the 1.12.2 tag and built the Docker images using `make docker.pilot docker.proxyv2 docker.operator docker.install-cni`. I pushed those to a private container registry and changed the `hub` in the global settings of my Istio installation. If you run with CNI you will also need the changes from #37166.

To get IPv4 working I had to add the following config as well (Helm format):

```yaml
global:
  proxy:
    excludeIPRanges: '0.0.0.0/0'
```
